app/blog/posts/good-python-code.mdx

```diff
@@ -0,0 +1,333 @@
+    writer.writerows(data)
+```
+
+After spending 30 minutes testing and fixing some bugs with this version, I discovered to my dismay that Pandas had the native `to_csv` classmethod to write to csv and that I could generate a Dataframe from a list of objects as seen below.
+
+```python
+import pandas as pd
+
+data = [{"key1": 2, "key4": 10}, {"key3": 3}, {"key4": 4}]
+df = pd.DataFrame(data)
+df.to_csv("output.csv", index=False)
+```
+
+What's beautiful about using a pandas dataframe is that now, you get all this beautiful added functionality like 
+
+- Generating it as a markdown table? - just use `df.to_markdown()`
+- Want to get a dictionary with the keys of each row? - just use `df.to_dict()`
+- Want to get a json formatted string? - just use `df.to_json()`
+
+That's a huge reduction in the potential issues with my code because I'm now using a library method that other people have spent time and effort to write and test. Standard libraries are also well supported across the ecosystem, allowing you to take advantage of other integrations down the line (Eg. LanceDB supporting Pydantic Models )
+
+## Write Simpler Functions
+
+I think that there are three big benefits to writing simpler functions that don't mutate state 
+
+1. They're easier to reason about since they're much smaller in size 
+2. They're easier to test because we can mock the inputs and assert on the outputs 
+3. They're easier to refactor because we can swap out different components easily
+
+I had a pretty complex problem to solve recently with some code - which was to take in a dataset of rows with some text and then embed every sentence inside it. I took some time and wrote an initial draft that looked something like this
+
+``` python
+def get_dataset_batches(data, dataset_mapping: dict[str, int], batch_size=100):
+    """
+    In this case, dataset_mapping maps a sentence to a
+    id that we can use to identify it by. This is an
+    empty dictionary by default
+    """
+    batch = []
+    for row in data:
+        s1, s2 = data["text"]
+        if s1 not in dataset_mapping:
+            dataset_mapping[s1] = len(dataset_mapping)
+            batch.append(s1)
+            if len(batch) == batch_size:
+                yield batch
+                batch = []
+
+        if s2 not in dataset_mapping:
+            dataset_mapping[s2] = len(dataset_mapping)
+            batch.append(s1)
+        if len(batch) == batch_size:
+            yield batch
+            batch = []
+
+    if batch:
+        yield batch
+```
+
+Instead of doing this, a better method might be to break up our function into the following few smaller functions as seen below.
+
+```python
+def get_unique_sentences(data):
+    seen = set()
+    for row in data:
+        s1, s2 = data["text"]
+        if s1 not in seen:
+            seen.add(s1)
+            yield s1
+
+        if s2 not in seen:
+            seen.add(s1)
+            yield s2
+
+
+def get_sentence_to_id_mapping(sentences: List[str]):
```

Comment:
looking again sn't this just 

`{s:i for i, s in enumerate(sentence)}`

Reply:
You are indeed right haha

Reply:
Changed it to

```
def get_sentence_to_id_mapping(sentences: List[str]):
    return {s:i for i, s in enumerate(sentence)}
```